auth/constructor.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"fmt"
+
+	"github.com/xmidt-org/themis/xlog"
+	"github.com/xmidt-org/webpa-common/basculechecks"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	gokitprometheus "github.com/go-kit/kit/metrics/prometheus"
+	"github.com/justinas/alice"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/bascule/basculehttp"
+	"github.com/xmidt-org/bascule/key"
+	"github.com/xmidt-org/webpa-common/basculemetrics"
+	"go.uber.org/fx"
+)
+
+type PrimaryCOptionsIn struct {
+	fx.In
+	Logger  log.Logger
+	Options []basculehttp.COption `group:"primary_bascule_constructor_options"`
+}
+
+type PrimaryTokenFactoryIn struct {
+	fx.In
+	Logger       log.Logger
+	DefaultKeyID string         `name:"primary_bearer_default_kid"`
+	Resolver     key.Resolver   `name:"primary_bearer_key_resolver" optional:"true"`
+	Leeway       bascule.Leeway `name:"primary_bearer_leeway"`
+}
+
+type PrimaryBasculeMetricListenerIn struct {
+	fx.In
+	Listener *basculemetrics.MetricListener `name:"primary_bascule_metric_listener"`
+}
+
+func ProvidePrimaryBasculeConstructor() fx.Option {
+	return fx.Provide(
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in PrimaryBasculeMetricListenerIn) basculehttp.COption {
+				return basculehttp.WithCErrorResponseFunc(in.Listener.OnErrorResponse)
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in PrimaryTokenFactoryIn) basculehttp.COption {
+				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "providing token factory option", "server", "primary")
+				if in.Resolver == nil {
+					in.Logger.Log(level.Key(), level.WarnValue(), xlog.MessageKey(), "providing nil token factory option as resolver was not defined", "server", "primary")
+					return nil
+				}
+				return basculehttp.WithTokenFactory("Bearer", basculehttp.BearerTokenFactory{
+					DefaultKeyId: in.DefaultKeyID,
+					Resolver:     in.Resolver,
+					Parser:       bascule.DefaultJWTParser,
+					Leeway:       in.Leeway,
+				})
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in primaryProfileIn) (basculehttp.COption, error) {
+				if in.Profile == nil {
+					return nil, nil
+				}
+				basicTokenFactory, factoryErr := basculehttp.NewBasicTokenFactoryFromList(in.Profile.Basic)
+				if factoryErr != nil {
+					return nil, factoryErr
+				}
+				return basculehttp.WithTokenFactory("Basic", basicTokenFactory), nil
+			},
+		},
+		fx.Annotated{
+			Name: "primary_alice_constructor",
+			Target: func(in PrimaryCOptionsIn) alice.Constructor {
+				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "providing alice constructor from bascule constructor options", "server", "primary")
+				if anyNil(in.Options) {
```

Comment:
Is it possible for any of these options to be nil? I think it depends on if nil values are added to a group.  If nil values aren't added, we're good.  But if we do somehow get a nil option, it seems like that's a problem since we want to build a constructor.  Why not just return an error here rather than continuing to try to build middleware that won't work?

Reply:
If any of the options are nil, it means that option was optional and so we don't build a constructor. If there were any legitimate errors building those options, this Target wouldn't get called.

Tr1d1um and Scytale both don't require any of these configs but fail with key not supported. Let's match that behavior?

Reply:
If an option is optional, shouldn't we still build a constructor?